Move lru cache from inside of `_encode_host` to outside (#1348)
aio-libs/yarl: docs/api.rst

Cache control
-------------

> IDNA conversion and host encoding encoding are quite expensive operations,

Author: ```suggestion
IDNA conversion and host encoding are quite expensive operations,
```

that's why the ``yarl`` library caches these calls by storing results in the
global LRU cache.

.. function:: cache_clear()

   Clear IDNA and host encoding cache.

.. function:: cache_info()

>    Return a dictionary with ``"idna_encode"``, ``"idna_decode"``, ``"ip_address"``,

Author: ```suggestion
   Return a dictionary with ``"idna_encode"``, ``"idna_decode"``,
```

>    ``"host_validate"``, and ``"encode_host"`` keys, each value

Author: ```suggestion
    and ``"encode_host"`` keys, each value
```

IDNA conversion, host validation, and IP Address parsing used for host
encoding are quite expensive operations, that's why the ``yarl``

> library caches these calls by storing last ``512`` results in the

Author: ```suggestion
library caches these calls by storing last ``256`` results in the
```

> global LRU cache.

Author: ```suggestion
global LRU cache. For encoding hosts, the last ``512`` results are stored.
```

      >>> yarl.cache_info()
      {'idna_encode': CacheInfo(hits=5, misses=5, maxsize=256, currsize=5),
       'idna_decode': CacheInfo(hits=24, misses=15, maxsize=256, currsize=15),
       'encode_host': CacheInfo(hits=0, misses=0, maxsize=512, currsize=0)}

   .. versionchanged:: 1.16

      ``ip_address``, and ``host_validate``
      are deprecated in favor of a single ``encode_host`` cache.

> .. function:: cache_configure(*, idna_encode_size=256, idna_decode_size=256, encode_host=512)

Author: ```suggestion
.. function:: cache_configure(*, idna_encode_size=256, idna_decode_size=256, encode_host_size=512)
```

>    Set the IP Address, host validation, and IDNA encode, host encode and

Author: ```suggestion
   Set the IDNA encode, IDNA decode, and host encode
```

>    decode cache sizes.

Author: ```suggestion
   cache sizes.
```
